pyccl/base.py

```diff
@@ -298,3 +299,332 @@ def increment(self):
+                abstracts = set(cls.__abstractmethods__) - set(linked)
+                cls.__abstractmethods__ = frozenset(abstracts)
+
+        return cls
+
+
+class CCLObject(metaclass=CCLObjectMeta):
+    """Base for CCL objects.
+
+    All CCL objects inherit ``__eq__`` and ``__hash__`` methods from here.
+    Both methods rely on ``__repr__`` uniqueness. This aims to homogenize
+    equivalence checking, and to standardize the use of hash.
+
+    Overview
+    --------
+    ``CCLObjects`` inherit ``__hash__``, which consistently hashes the
+    representation string. They also inherit ``__eq__`` which checks for
+    representation equivalence.
+
+    In the implemented scheme, each ``CCLObject`` may have its own, specialized
+    ``__repr__`` method overloaded. Object representations have to be unique
+    for equivalent objects. If no ``__repr__`` is provided, the default from
+    ``object`` is used.
+
+    Mutation
+    --------
+    ``CCLObjects`` are by default immutable. This aims to provide a failsafe
+    mechanism, where, changing attributes has to trigger a re-computation
+    of something else inside of the instance, rather than simply doing a value
+    change.
+
+    This immutability mechanism can be safely bypassed if a subclass defines an
+    ``update_parameters`` method. ``CCLObjects`` temporarily unlock whenever
+    this method is called.
+
+    Internal State vs. Mutation
+    ---------------------------
+    Other methods that use ``setattr`` can only do that if they are decorated
+    with ``@unlock_instance`` or if the particular code block that makes the
+    change is enclosed within the ``UnlockInstance`` context manager.
+    If neither is provided, an exception is raised.
+
+    If such methods only change the instance's internal state, the decorator
+    may be called with ``@unlock_instance(mutate=False)`` (or equivalently
+    for the context manager ``UnlockInstance(..., mutate=False)``). Otherwise,
+    the instance is assumed to have mutated.
+    """
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+
+        # 1. Store the signature of the constructor on import.
+        cls.__signature__ = signature(cls.__init__)
+
+        # 2. Assign the arguments of the constructor if needed.
+        if (hasattr(cls, "__init_attrs__")
+                and cls.__init_attrs__
+                and "__init__" in vars(cls)):
+            cls.__init__ = auto_assign(cls.__init__, sig=cls.__signature__)
+
+        # 3. Replace repr (if implemented) with its cached version.
+        if "__repr__" in vars(cls):
+            # If the class defines a custom `__repr__`, this will be the new
+            # `_repr` (which is cached). Decorator `cached_property` requires
+            # that `__set_name__` is called on it.
+            bmethod = functools.cached_property(cls.__repr__)
+            cls._repr = bmethod
+            bmethod.__set_name__(cls, "_repr")
```

Comment:
Why do we need both of the two lines above? Wouldn't `cls._repr` already assign a method called `_repr` to `cls`?

Reply:
Python doesn't bind methods to an object from outside `__init__`. There are two kinds of objects: functions, and bound methods. If we were to omit L560, the function wouldn't become a bound method, and the implicit first argument (`self`) wouldn't refer to an instance of that class. So, in this way, you're telling the class dict that what you give it is actually a bound method, not a function.